### ISSUE TITLE
CMake doesn't generate the paho-mqttpp3.lib on Microsoft Windows

### ISSUE THREAD

**Issue by guilhermeferreira** [eclipse/paho.mqtt.cpp#117]
Both the unit test binary (**mqttpp-unittest**) and the samples binaries require the **paho-mqttpp3.lib**. However, the CMake scripts are creating only the **paho-mqttpp3.dll**.

```
Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

C:\Temp\paho.mqtt.cpp\build_cmake>"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" x64

C:\Temp\paho.mqtt.cpp\build_cmake>cmake -G "NMake Makefiles" -DCMAKE_INSTALL_PREFIX="C:\Temp\paho-cpp" -DPAHO_MQTT_C_PATH="C:\Temp\paho-c" -DPAHO_WITH_SSL=FALSE -DPAHO_BUILD_DOCUMENTATION=FALSE -DCMAKE_CXX_FLAGS_RELEASE="-IC:\Temp\GTest" -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=FALSE ..
-- The CXX compiler identification is MSVC 19.0.23918.0
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/amd64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/amd64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is MSVC 19.0.23918.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/amd64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/amd64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Temp/paho.mqtt.cpp/build_cmake

C:\Temp\paho.mqtt.cpp\build_cmake>nmake

Microsoft (R) Program Maintenance Utility Version 14.00.23918.0
Copyright (C) Microsoft Corporation.  All rights reserved.

Scanning dependencies of target common_obj
[  6%] Building CXX object src/CMakeFiles/common_obj.dir/async_client.cpp.obj
async_client.cpp
[ 13%] Building CXX object src/CMakeFiles/common_obj.dir/client.cpp.obj
client.cpp
[ 20%] Building CXX object src/CMakeFiles/common_obj.dir/disconnect_options.cpp.obj
disconnect_options.cpp
[ 26%] Building CXX object src/CMakeFiles/common_obj.dir/iclient_persistence.cpp.obj
iclient_persistence.cpp
[ 33%] Building CXX object src/CMakeFiles/common_obj.dir/message.cpp.obj
message.cpp
[ 40%] Building CXX object src/CMakeFiles/common_obj.dir/response_options.cpp.obj
response_options.cpp
[ 46%] Building CXX object src/CMakeFiles/common_obj.dir/ssl_options.cpp.obj
ssl_options.cpp
[ 53%] Building CXX object src/CMakeFiles/common_obj.dir/string_collection.cpp.obj
string_collection.cpp
[ 60%] Building CXX object src/CMakeFiles/common_obj.dir/token.cpp.obj
token.cpp
[ 66%] Building CXX object src/CMakeFiles/common_obj.dir/topic.cpp.obj
topic.cpp
[ 73%] Building CXX object src/CMakeFiles/common_obj.dir/connect_options.cpp.obj
connect_options.cpp
[ 80%] Building CXX object src/CMakeFiles/common_obj.dir/will_options.cpp.obj
will_options.cpp
[ 80%] Built target common_obj
Scanning dependencies of target paho-mqttpp3
[ 86%] Linking CXX shared library paho-mqttpp3.dll
[ 86%] Built target paho-mqttpp3
Scanning dependencies of target mqttpp-unittest
[ 93%] Building CXX object test/unit/CMakeFiles/mqttpp-unittest.dir/test.cpp.obj
test.cpp
NMAKE : fatal error U1073: don't know how to make 'src\paho-mqttpp3.lib'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.

C:\Temp\paho.mqtt.cpp\build_cmake>dir src
 Volume in drive C is OSDisk
 Volume Serial Number is C2D6-0DCE

 Directory of C:\Temp\paho.mqtt.cpp\build_cmake\src

05/31/2017  04:35 PM    <DIR>          .
05/31/2017  04:35 PM    <DIR>          ..
05/31/2017  04:34 PM    <DIR>          CMakeFiles
05/31/2017  04:34 PM             1,449 cmake_install.cmake
05/31/2017  04:34 PM            25,405 Makefile
05/31/2017  04:34 PM    <DIR>          mqtt
05/31/2017  04:35 PM           594,944 paho-mqttpp3.dll
05/31/2017  04:35 PM               381 paho-mqttpp3.dll.manifest
               4 File(s)        622,179 bytes
               4 Dir(s)  809,278,922,752 bytes free
```

**Comment by guilhermeferreira** [comment]
According to this [Stackoverflow's question](https://stackoverflow.com/questions/7614286/how-do-i-get-cmake-to-create-a-dll-and-its-matching-lib-file), the **paho-mqttpp3.lib** is not created because there are no symbols exported. Then, I used the `DLLExport` macro from Paho MQTT C to test:

```
class DLLExport async_client : public virtual iasync_client
{
    // ...
};
```

And it forces the Visual Studio to create the  **paho-mqttpp3.lib**:

```
C:\Temp\paho.mqtt.cpp\build_cmake>dir src
 Volume in drive C is OSDisk
 Volume Serial Number is C2D6-0DCE

 Directory of C:\Temp\paho.mqtt.cpp\build_cmake\src

06/01/2017  11:54 AM    <DIR>          .
06/01/2017  11:54 AM    <DIR>          ..
06/01/2017  11:54 AM    <DIR>          CMakeFiles
05/31/2017  06:23 PM             1,449 cmake_install.cmake
05/31/2017  06:23 PM            25,405 Makefile
06/01/2017  11:54 AM    <DIR>          mqtt
05/31/2017  06:24 PM           639,488 paho-mqttpp3.dll
05/31/2017  06:24 PM               381 paho-mqttpp3.dll.manifest
05/31/2017  06:24 PM            94,568 paho-mqttpp3.exp
05/31/2017  06:24 PM           156,904 paho-mqttpp3.lib
               6 File(s)        918,195 bytes
               4 Dir(s)  809,452,789,760 bytes free

```

**Comment by fpagliughi** [comment]
Did you try to build any executables (like the samples) after you did this? I'm getting a .lib file (although it's only 32kB) and getting all sorts of errors and warnings now.

**Comment by guilhermeferreira** [comment]
I guess the `DLLExport` is required in every class/symbol the samples (or other binary) will use. Otherwise, you will get a lot of "undefined reference" errors. Also, I'm struggling with other errors, like library paths and machine architecture mismatch. The things get complicated because I don't have too much experience with NMake.

**Comment by fpagliughi** [comment]
I went through and tried adding it to all the classes, but then started having errors from static data, so then tried to figure that out. 

I also got lots and lots of C4251 warnings. This seems to be mostly (if not all) from things that Microsoft tells you to ignore, but I can't really be sure if they are all safe to ignore without reading through each and every of them, and I don't really feel like doing that!

What's worrisome, though, is a couple of issues raised by the documentation for the warning:
https://msdn.microsoft.com/query/dev14.query?appId=Dev14IDEF1&l=EN-US&k=k(C4251);k(vs.output)&rd=true

 - No inlined methods of your class can modify static data.
 - No inlined methods of your class use CRT functions or other library functions use static data 

I don't know if I can make this promise, or that I can remember it in the future! 

By this, I would claim that Windows DLL's are not fully compliant with C++11, and that is the one thing I have tried to stick with, and require, in this library.

So, basically, if the static library CMake build works for Windows, I am content that release 1.0 doesn't support DLL's. Then we can make an issue to be addressed in the next release, if ever.

**Comment by fpagliughi** [comment]
Can we use CMake to just build a static .lib file on Windows?

**Comment by guilhermeferreira** [comment]
> So, basically, if the static library CMake build works for Windows,

@fpagliughi yes, it works. Clients can link against the **paho-mqttpp3-static.lib** instead of the pair  **paho-mqttpp3.lib** / **paho-mqttpp3.dll**. 

> I am content that release 1.0 doesn't support DLL's. 

We must state somewhere (documentation? enforce through the CMakeLists.txt?) that the current version does not support DLLs.

> Then we can make an issue to be addressed in the next release, if ever.

Perfect.

> Can we use CMake to just build a static .lib file on Windows?

Yes, it already does. If the option `PAHO_BUILD_STATIC` is enabled , the compilation produces the **paho-mqttpp3-static.lib** file. 

```
cmake -G "NMake Makefiles" -DCMAKE_INSTALL_PREFIX="C:\Temp\paho-cpp" -DPAHO_MQTT_C_PATH="C:\Temp\paho-c" -DPAHO_WITH_SSL=FALSE -DPAHO_BUILD_DOCUMENTATION=FALSE -DPAHO_BUILD_STATIC=TRUE -DGTEST_PATH="C:\Temp\GTest" -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=TRUE ..
```

**Comment by fpagliughi** [comment]
If you ask for a static build, does it still try to build the DLL?

**Comment by guilhermeferreira** [comment]
> If you ask for a static build, does it still try to build the DLL?

Yes, @fpagliughi. Currently, the static library creation is optional, the DLL is mandatory.

**Comment by fpagliughi** [comment]
Fixed in the 'develop' branch. Windows can now build static and shared lib.
